### ISSUE TITLE
prune away some of the annotations before the release that are not useful for users

### ISSUE THREAD

**Issue by nicolevasilevsky** [monarch-initiative/mondo#4346]
@joeflack4 to do: (_update: I merged PR #4347. This is no longer blocked._)

Prune away some of the annotations before the release that are not useful for users (@matentzn will show Joe how to do ROBOT remove)
- Exclude from QC check
- MONDO:preferredExternal
- confidence
- has_synonym_type clingen_preferred (this is an annotation on synonyms)

**Comment by matentzn** [comment]
- [ ] Familiarise yourself with the Makefile dependency chain for [mondo](https://github.com/monarch-initiative/mondo/blob/master/src/ontology/Makefile#L234) by tracing the dependencies up to `$(SRC)`.
- [ ] Extend the owltools command [for filtered.owl](https://github.com/monarch-initiative/mondo/blob/master/src/ontology/Makefile#L212) with an additional `ROBOT remove` command right afterwards. Bonus: replace the owltools command altogether with a suitable ROBOT command if you can figure out how (no sweat).
- [ ]  Create a config file in a suitable location, maybe `src/ontology/config/remove-annotations-before-release.txt`
- [ ] Add all annotation properties @nicolevasilevsky tell you to that file
- [ ] The ROBOT remove command should simply remove all of them with default settings.
- [ ] Run the command with and without the remove, compare the two resulting files and ensure that nothing important was lost.

**Comment by nicolevasilevsky** [comment]
These are the annotation properties:
- clingen_preferred, IRI: http://purl.obolibrary.org/obo/mondo#clingen_preferred
- excluded from QC check, IRI: http://purl.obolibrary.org/obo/mondo#excluded_from_qc_check
- confidence, IRI: http://purl.obolibrary.org/obo/mondo#confidence
- preferred external ID, IRI: http://purl.obolibrary.org/obo/mondo#preferred_external_id

**Comment by joeflack4** [comment]
@nicolevasilevsky @matentzn Thanks guys. I found some instructions on how to install `owltools`. I'll do that and then try the rest of these steps.

**Comment by matentzn** [comment]
Is there a technical reason you avoid to use docker? Eventually you will need to use it unless you want to rewrite dozens of commands in other parts of the Makefile Process :)

**Comment by joeflack4** [comment]
I use docker. But what would I be using it for in this case?

**Comment by matentzn** [comment]
why do you install owltools and robot locally when you can invoke them through odk? The sh run.sh wraps all subsequent commands inside the odk docker container..

**Comment by joeflack4** [comment]
I guess I don't have a strong preference, but I guess it is slightly less complex to run them directly (for me at least). I haven't fully internalized the entire Mondo stack in my head; I've been learning things here and there in a piecemeal fashion. That's probably why I'm doing it a little differently.

I could have sworn I ran robot through docker in the past, though. But lately I've been running it directly. I might have accidentally deleted it from my docker dashboard. 

The owltools installation steps look really simple. But if for some reason I run into some trouble, I'll go with the docker method.

**Comment by joeflack4** [comment]
## Details
_(skip to the bottom for TLDR)_

I wasn't sure which arguments to use, but I might have figured it out.

I looked at `robot remove -h` and could not find mention of "annotations". I did see mention of axioms, terms, and relationships, though, so I guessed that "annotations" would fall under one of those.

```
robot remove -h
usage: robot remove --input <file> --output <file>
 -a,--axioms <arg>               filter only for given axiom types
    --add-prefix <arg>           add prefix 'foo: http://bar' to the
                                 output
    --add-prefixes <arg>         add JSON-LD prefixes to the output
    --base-iri <arg>             specify a base namespace
    --catalog <arg>              use catalog from provided file
 -E,--exclude-terms <arg>        set of terms in text file to exclude from
                                 removal
 -e,--exclude-term <arg>         term to exclude from removal
 -h,--help                       print usage information
 -I,--input-iri <arg>            load ontology from an IRI
 -i,--input <arg>                load ontology from a file
    --include-term <arg>         include term
 -N,--include-terms <arg>        set of terms in file to force include
 -n,--include-term <arg>         term to force include
 -noprefixes                     do not use default prefixes
 -o,--output <arg>               save ontology to a file
 -P,--prefixes <arg>             use prefixes from JSON-LD file
 -p,--preserve-structure <arg>   if false, do not preserve hierarchical
                                 relationships
 -r,--trim <arg>                 if true, remove axioms containing any
                                 selected object
 -S,--signature <arg>            if true, remove axioms with any selected
                                 entity in their signature
 -s,--select <arg>               select a set of terms based on relations
    --strict                     use strict parsing when loading an
                                 ontology
 -T,--term-file <arg>            load terms from a file
 -t,--term <arg>                 term to remove
 -V,--version                    print version information
 -v,--verbose                    increased logging
 -vv,--very-verbose              high logging
 -vvv,--very-very-verbose        maximum logging, including stack traces
 -x,--xml-entities               use entity substitution with ontology XML
                                 output
```

I decided to look through some of the `robot` source code to see if I could find if any of these arguments were related to annotations (don't worry; didn't take me much time). It looked like [`--select` seemed most related to annotations](https://github.com/ontodev/robot/blob/master/robot-command/src/main/java/org/obolibrary/robot/RemoveCommand.java#L122), but when I tried it, I didn't get good results. You can see the makefile target/recipe/command I made below, and some comments regarding the error that I got. I tried both `--select ANNOTATION_A ANNOTATION_B ...` AND `--select ANNOTATION_A --select ANNOTATION_B` syntax, but neither seemed to work for me.

```make
# X: A single arg on a single --select worked. multiple args on a single --select wouldn't even run the command. multiple --selects produced err
# ... I'm confused by the error because it seems to be expecting a pattern, but `robot remove -h` `--select` help text doesn't indicate anything about patterns.
#
# make filtered3.owl; diff -u filtered.owl filtered3.owl > filtered3.diff; beep
# robot --catalog catalog-v001.xml remove --input filtered.owl --output filtered3.owl --select http://purl.obolibrary.org/obo/mondo#CLINGEN_PREFERRED --select http://purl.obolibrary.org/obo/mondo#excluded_from_qc_check --select http://purl.obolibrary.org/obo/mondo#confidence --select MONDO:preferredExternal
# ^[[A^[[A^[[B2022-01-24 18:59:57,016 ERROR org.obolibrary.robot.RelatedObjectsHelper - Prefix 'http' is not a loaded prefix and will be ignored
# 2022-01-24 18:59:57,022 ERROR org.obolibrary.robot.RelatedObjectsHelper - Prefix 'http' is not a loaded prefix and will be ignored
# 2022-01-24 18:59:57,029 ERROR org.obolibrary.robot.RelatedObjectsHelper - Prefix 'http' is not a loaded prefix and will be ignored
# INVALID IRI PATTERN ERROR the pattern 'http://purl.obolibrary.org/obo/MONDO_preferredExternal' must contain at least one wildcard character.
# For details see: http://robot.obolibrary.org/errors#invalid-iri-pattern-error
# Invalid IRI Pattern Error
# When matching an IRI by pattern, the pattern should contain one or more wildcard characters (* and/or ?) or should be a regex pattern preceded by ~. If you wish to match a full IRI to remove or filter, use the --term option.
filtered3.owl: filtered.owl
	$(ROBOT) remove --input $< --output $@ --select http://purl.obolibrary.org/obo/mondo#CLINGEN_PREFERRED --select http://purl.obolibrary.org/obo/mondo#excluded_from_qc_check --select http://purl.obolibrary.org/obo/mondo#confidence --select MONDO:preferredExternal
```

I also tried `--axioms`, but was no good:
```make
# X: Just tried a single item here, but got error anyway:
# AXIOM TYPE ERROR http://purl.obolibrary.org/obo/mondo#clingen_preferred is not a valid axiom type
# For details see: http://robot.obolibrary.org/errors#axiom-type-error
filtered5.owl: filtered.owl
	$(ROBOT) remove --input $< --output $@ --axioms http://purl.obolibrary.org/obo/mondo#excluded_from_qc_check
```

However, using the _terms_ argument for removing annotations seems like it might have worked. This makes sense, because Nico had mentioned creating a "config file", and the only one of the CLI arguments that accepted such a file was `--term-file`. I had tried this first but didn't get a good result; I was searching for `clingen_preferred` but saw it hadn't been removed. It turned out I needed to change it to uppercase.

Out of the 4 total annotations from Nicole, these look like they might have been adequately removed:
- http://purl.obolibrary.org/obo/mondo#CLINGEN_PREFERRED
- http://purl.obolibrary.org/obo/mondo#clingen_preferred
- http://purl.obolibrary.org/obo/mondo#excluded_from_qc_check

For `clingen_preferred`, I had to add an uppercase entry. Actually, all instances of this annotation were capitalized except for 1. 

This one didn't appear in input, but I imagine that if the ones above worked, it would also have worked:
- http://purl.obolibrary.org/obo/mondo#confidence

@nicolevasilevsky @matentzn The one I'm having the most difficulty and could use help with is:
- MONDO:preferredExternal

I tried this in both prefix format (`MONDO:preferredExternal`) and IRI format (http://purl.obolibrary.org/obo/mondo#preferredExternal), but neither seem to be working. For the record, I am only seeing this in the prefix format within the `filtered.owl`.

As I had suspected, the output of these were both the same:
```make
filtered2.owl: filtered.owl
	$(ROBOT) remove --input $< --output $@ --term-file config/remove-annotations-before-release.txt

filtered4.owl: filtered.owl
	$(ROBOT) remove --input $< --output $@ --term http://purl.obolibrary.org/obo/mondo#CLINGEN_PREFERRED --term http://purl.obolibrary.org/obo/mondo#clingen_preferred --term http://purl.obolibrary.org/obo/mondo#excluded_from_qc_check --term http://purl.obolibrary.org/obo/mondo#confidence --term MONDO:preferredExternal
```

The `--term-file` option is consistent with what Nico requested, so I kept that one.

I then changed the name of the original `filtered.owl` recipe to `filtered-step1.owl`, and changed the name of my `filtered2.owl` recipe above to `filtered.owl`. The reason I did this is because there are other makefile recipes that are using `filtered.owl`, and I think that they would prefer to use the recipe that I just made instead.

@matentzn If you'd like, I can combine these two recipes together, maybe something like this (I don't know how to do it all in `robot` alone)? Let me know what you think:
```make
filtered.owl: filtered.obo skos.ttl
	owltools --use-catalog $< $(REMOVE_DECLARATIONS) skos.ttl --merge-support-ontologies -o filtered-temp.owl
	$(ROBOT) remove --input filtered-temp.owl --output $@ --term-file config/remove-annotations-before-release.txt
```

## TLDR / summary
Looks like things might have worked, hopefully.

I've attached the output of `diff -u filtered.owl filtered2.owl > filtered2.diff`:
[filtered2.diff.zip](https://github.com/monarch-initiative/mondo/files/7930397/filtered2.diff.zip)

Also, here's the `filtered2.owl` that was generated. It looks like what I did worked, but I'm not entirely sure if it did exactly what we want or if there were any unintended side effects. @nicolevasilevsky @matentzn if you guys could review, I would appreciate:
https://drive.google.com/file/d/1E9XPV6oi70ChC_yGOjkvW1MxesD4lOdP/view?usp=sharing

I went ahead and submitted a PR: https://github.com/monarch-initiative/mondo/pull/4357

**Comment by matentzn** [comment]
Looks like good work @joeflack4 I think the way you do it is mostly right. We will review both your PRs together in the QC call in depth, but from what I can see, you mostly nailed this one here!

**Comment by joeflack4** [comment]
Thanks Nico for the explanation. Sorry about my internet being down last Friday. I'll talk with you more about this during one of our calls this week or next!
